contracts/vortex/CarbonVortex.sol

```diff
@@ -138,109 +345,502 @@ contract CarbonVortex is ICarbonVortex, Upgradeable, ReentrancyGuardUpgradeable,
     function execute(Token[] calldata tokens) external nonReentrant validateTokens(tokens) {
         uint256 len = tokens.length;
 
-        // allocate balances array for the tokens
-        uint256[] memory balances = new uint256[](len);
+        // allocate array for the fee amounts for the tokens
+        uint256[] memory feeAmounts = new uint256[](len);
         // allocate array for the reward amounts for caller
         uint256[] memory rewardAmounts = new uint256[](len);
         // cache rewardsPPM to save gas
         uint256 rewardsPercentage = _rewardsPPM;
 
-        // withdraw fees, load balances and reward amounts
+        // withdraw fees from carbon, vault and old vortex
         for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
             // withdraw token fees
-            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
-            // get token balance
-            balances[i] = tokens[i].balanceOf(address(this));
+            uint256 totalFeeAmount = 0;
+            if (address(_carbonController) != address(0)) {
```

Comment:
Perhaps these checked can be cached into a boolean outside of the loop

Reply:
fixed